validator-api/src/epoch_operations/mod.rs

```diff
@@ -90,44 +99,113 @@ impl RewardedSetUpdater {
 
     async fn determine_layers(
         &self,
-        rewarded_set: &[MixNodeDetails],
+        rewarded_set: &mut Vec<MixNodeDetails>,
     ) -> Result<(Vec<LayerAssignment>, HashMap<String, Layer>), RewardingError> {
         let mut families_in_layer: HashMap<String, Layer> = HashMap::new();
-        let mut assignments = vec![];
-        let mut layer_assignments: HashMap<Layer, f32> = HashMap::new();
-        let mut rng = OsRng;
-        let layers = vec![Layer::One, Layer::Two, Layer::Three];
+        let mut assignments = Vec::with_capacity(rewarded_set.len());
+        let target_layer_count = rewarded_set.len() / 3;
 
         let mix_to_family = self.validator_cache.mix_to_family().await.to_vec();
 
         let mix_to_family = mix_to_family
             .into_iter()
             .collect::<HashMap<IdentityKey, FamilyHead>>();
 
-        for mix in rewarded_set {
-            let family = mix_to_family.get(&mix.bond_information.identity().to_owned());
-            // Get layer already assigned to nodes family, if any
-            let family_layer = family.and_then(|h| families_in_layer.get(h.identity()));
+        let mut nodes_in_families = vec![];
+        let mut regular_nodes = Vec::with_capacity(rewarded_set.len());
 
-            // Same node families are always assigned to the same layer, otherwise layer selected by a random weighted choice
-            let layer = if let Some(layer) = family_layer {
-                layer.to_owned()
+        for (idx, node) in rewarded_set.drain(..).enumerate() {
```

Comment:
```suggestion
        for (idx, node) in rewarded_set.into_iter().enumerate() {
```
so that we could get rid of `&mut` in the argument?